### DISCUSSION TITLE
/security/authentication.html

### DISCUSSION THREAD

**Discussion by giscus** [goravel/goravel#293] [Announcements]
# /security/authentication.html

[[toc]] 简介 用户认证是 Web 应用中不可或缺的功能，Goravel 的 facades.Auth() 模块提供 JWT 功能的支持。 配置 可以在 config/auth.go 文件中配置默认守卫与多个守卫，以便在应用中进行不同用户身份的切换。 可以在 config/jwt.go 文件中配置 JWT 的相关参数，如秘钥、有效时长、可刷新时长...

https://www.goravel.dev/zh/security/authentication.html

**Comment by okami-chen** [comment]
黑名单功能怎么支持？多设备怎么支持？

**Reply by hwbrzzl** [reply]
目前可以自己维护一个 Token 表实现。

**Comment by agomezguru** [comment]
Hello there:
Before anything, thanks a lot by the great work your team are doing. In Laravel when I need to protect all my routes to be accessed only by authorized users, I do in this way:

Route::middleware(['auth','checkstatus'])->group(function () {

  /**
   * Dashboard
   */
  Route::group(['middleware' => ['permission:dashboard|dashboard.own']], function () {
    Route::get('dashboard/view-rol-details', 'DashboardController@viewRolDetails');
    Route::get('dashboard', 'DashboardController@dashboard')->name('dashboard');
  });
});

How to accomplish the same thing with Goravel?

**Reply by hwbrzzl** [reply]
There is an example: https://github.com/goravel/example/blob/392c747cfc868fe5e04be1859935b76b695236fe/routes/web.go#L35

**Comment by saqijaan** [comment]
How can I register custom Auth Guard like in laravel?

**Reply by hwbrzzl** [reply]
Here: https://github.com/goravel/goravel/blob/3b9e0e2b1ab0cb0aa0fcb8bd3958d3b2a46460de/config/auth.go#L31

**Reply by saqijaan** [reply]
Thats okay but how to define class/struct. Like in laravel we used to do Auth::extend to register gaurd and then define config for that guard in auth.php

My use case is I've an existing application which is using Sanctum token for authentication. Now Goravel Supports only JWT. My app is for IOT devices and I don't want to break existing connections. What I want to do is create a custom Guard where I'll implement Sanctum token Parsing/Login etc with database table/model. 

For now I've overridden the Auth registration and modified it to support my use case. But What I want is for website things use JWT token and for IOT side use Sanctum token. 

Can you provide an example for that. It will be very helpful. Thanks. 

I like goravel very much and want to contribute it too but there is a very less incomplete documentation. 

You made a chesious thing bro.

**Reply by saqijaan** [reply]
Or we can say driver for new guard.

**Reply by hwbrzzl** [reply]
Yes, only support JWT for now, created an issue to implement this: https://github.com/goravel/goravel/issues/478, welcome to PR, it should not be complicated.

**Comment by backup-siz** [comment]
Hi, how can i make custom payload for register token?

**Reply by dev-aligh** [reply]
This is my question too

**Reply by hwbrzzl** [reply]
Could you introduce your requirement?

**Reply by dev-aligh** [reply]
For example i want to store name of user.

**Reply by hwbrzzl** [reply]
You can pass UserName to the `LoginUsingID` method to generate a token, but if so, the `User` method can't be used, because the `User` method gets the User model via the key, your key is UserName.

**Comment by backup-siz** [comment]
Another question is when we make a token with this code:

token, err := facades.Auth(ctx).LoginUsingID(1)

Where is the token stored in application?

**Reply by hwbrzzl** [reply]
Not store it in application, the token contains the ID, and the ID will be parsed via token.

**Comment by raydenka21** [comment]
https://laravel.com/docs/11.x/http-client goravel available ?

**Reply by hwbrzzl** [reply]
No for now, please find a third package.

**Comment by klphp** [comment]
```
type User struct {
  ID uint `gorm:"primaryKey"`
  Name string
}

var user models.User
user.ID = 1
user.Name= `zhang3`

token, err := facades.Auth(ctx).Login(&user)
```
解析后能获取user.Name吗

**Reply by hwbrzzl** [reply]
可以

**Reply by klphp** [reply]
当我使用facades.Auth(ctx).Guard(`admin`).User(&user)查询用户的时候，我发现默认是以user id查询用户数据。但有一个问题，就是说当我想返回用户数据的时候，实际上我没必要再一次查询，直接将结果返回给前端就可以了。实际应用场景中，我可能只想返回前端id,username。但facades.Auth(ctx).Guard(`admin`).User(&user)默认将所有字段都查出来。因此facades.Auth(ctx).Guard(`admin`).User(&user)是不是增加一个select更方便一些。或者有其它的方法可以直接屏蔽掉不想返回给前端的字段，如password。

**Reply by hwbrzzl** [reply]
增加 Select 方法有待探讨，返回给前端的数据像过滤字段的话，可以在 model 中自定义一个方法返回前端需要的字段。

**Comment by damon35868** [comment]
请问不同的guard可以设置不同的secret吗？

**Reply by hwbrzzl** [reply]
目前还不行，确实需要的话可以提个 issue。

**Reply by damon35868** [reply]
好的谢谢

**Comment by dcynsd** [comment]
请问能否支持 Payload 自定义内容？

**Reply by hwbrzzl** [reply]
目前不行
